Update Testing Framework docs for current version of Cadence (#213)
onflow/docs: versioned_docs/version-stable/cadence/testing-framework.mdx

Tests must be written in the form of a Cadence script.
A test script may contain testing functions that starts with the `test` prefix,

> a `setup` function that always runs before the tests,
> a `tearDown` function that always runs at the end of all test cases,
> a `beforeEach` function that runs before each test case,
> and an `afterEach` function that runs after each test case.
> All the above four functions are optional.

Reviewer: ```suggestion

Additionally, it can have below special functions:
- A `setup` function that always runs before the tests.
- A `tearDown` function that always runs at the end of all test cases.
- A `beforeEach` function that runs before each test case.
- An `afterEach` function that runs after each test case.

All the above four functions are optional.
```

```cadence
// A `setup` function that always runs before the rest of the test cases.
// Can be used to initialize things that would be used across the test cases.
// e.g: initialling a blockchain backend, initializing a contract, etc.
access(all) fun setup() {
}

// The `beforeEach` function runs before each test case. Can be used to perform
// some state cleanup before each test case, among other things.
access(all) fun beforeEach() {
}

// The `afterEach` function runs after each test case.  Can be used to perform
// some state cleanup after each test case, among other things.
access(all) fun afterEach() {
}

// Valid test functions start with the 'test' prefix.
access(all) fun testSomething() {
}

access(all) fun testAnotherThing() {
}

access(all) fun testMoreThings() {
}

> // Test functions cannot have any arguments or return values.

Reviewer: ```suggestion
// Invalid: Test functions cannot have any arguments or return values.
```

For example, a matcher that checks whether a given integer value is negative can be defined as follows:

```cadence
import Test

access(all) fun testExample() {
    let isNegative = Test.newMatcher(fun (_ value: Int): Bool {
        return value < 0
    })

    Test.expect(-15, isNegative)

>     // Alternatively, we can use `Test.assert` and the matcher's `test` function.

Reviewer: ```suggestion

    // Alternatively, we can use `Test.assert` and the matcher's `test` function.
```

```

The `haveElementCount` function returns a matcher that succeeds if the tested value is an array or dictionary,
and has the given number of elements.

```cadence
import Test

access(all) fun testExample() {
    let array: [String] = ["one", "two", "three"]

    Test.expect(array, Test.haveElementCount(3))

    let dictionary: {String: Int} = {"one": 1, "two": 2, "three": 3}

    Test.expect(dictionary, Test.haveElementCount(3))
}
```

### Test.contain

```cadence
fun contain(_ element: AnyStruct): Matcher
```

The `contain` function returns a matcher that succeeds if the tested value is an array that contains
a value that is equal to the given value, or the tested value is a dictionary
that contains an entry where the key is equal to the given value.

```cadence
access(all) fun testExample() {
    let array: [String] = ["one", "two", "three"]

    Test.expect(array, Test.contain("one"))

    let dictionary: {String: Int} = {"one": 1, "two": 2, "three": 3}

    Test.expect(dictionary, Test.contain("two"))
}
```

### Test.beSucceeded

```
fun beSucceeded(): Matcher
```

The `beSucceeded` function returns a new matcher that checks if the given test value is either

> a ScriptResult or TransactionResult and the ResultStatus is succeeded.

Reviewer: ```suggestion
a `ScriptResult` or `TransactionResult`, and the `ResultStatus` is succeeded.
```

Returns false in any other case.

```cadence
import Test

access(all) fun testExample() {
    let blockchain = Test.newEmulatorBlockchain()
    let result = blockchain.executeScript(
        "access(all) fun main(): Int {  return 2 + 3 }",
        []
    )

    Test.expect(result, Test.beSucceeded())
    Test.assertEqual(5, result.returnValue! as! Int)
}
```

### Test.beFailed

```cadence
fun beFailed(): Matcher
```

The `beFailed` function returns a new matcher that checks if the given test value is either

> a ScriptResult or TransactionResult and the ResultStatus is failed.

Reviewer: ```suggestion
a `ScriptResult` or `TransactionResult`, and the `ResultStatus` is failed.
```
